File: ruby/lib/enolib/elements/list.rb
# frozen_string_literal: true

module Enolib
  class List < ElementBase
    attr_reader :instruction

    def _instantiate_items(list)
      if list.has_key?(:mirror)
        _instantiate_items(list[:mirror])
      elsif list.has_key?(:extend)
        _instantiate_items(list[:extend]) + list[:items].map { |item| ListItem.new(@context, item, self) }
      elsif list.has_key?(:items)
        list[:items].map { |item| ListItem.new(@context, item, self) }
      else
        []
      end
    end

    def _items
      unless instance_variable_defined?(:@instantiated_items)
        @instantiated_items = _instantiate_items(@instruction)
      end

      @instantiated_items
    end

    def _untouched
      return @instruction unless instance_variable_defined?(:@touched)

      untouched_item = _items.find { |item| !item.instance_variable_defined?(:@touched) }

      untouched_item ? untouched_item.instruction : false
    end

    def items
      @touched = true

      _items
    end

    def length
      @touched = true

      _items.length
    end

    def optional_comment(loader = nil)
      loader = Proc.new if block_given?

      if loader
        _comment(loader, required: false)
      else
        raise ArgumentError.new('A loader block or Proc must be provided')
      end
    end

    def optional_string_values
      @touched = true

      _items.map(&:optional_string_value)
    end

    def optional_values(loader = nil)
      loader = Proc.new if block_given?

      @touched = true

      unless loader
        raise ArgumentError.new('A loader function must be provided')
      end

      _items.map { |item| item.optional_value(loader) }
    end

    def parent
      @parent || Section.new(@context, @instruction[:parent])
    end

    def required_string_values
      @touched = true

      _items.map(&:required_string_value)
    end

    def required_values(loader = nil)
      loader = Proc.new if block_given?

      @touched = true

      unless loader
        raise ArgumentError.new('A loader function must be provided')
      end

      _items.map { |item| item.required_value(loader) }
    end

    def to_s
      "#<Enolib::List key=#{@instruction[:key]} items=#{_items.length}>"
    end

    def touch
      @touched = true

      _items.each(&:touch)
    end
  end
end

File: python/enolib/elements/list.py
from . import list_item
from . import section
from .element_base import ElementBase


class List(ElementBase):
    def __repr__(self):
        return f"<class List key={self._instruction['key']} items={len(self._items())}>"

    def _instantiate_items(self, list):
        if 'mirror' in list:
            return self._instantiate_items(list['mirror'])
        elif 'extend' in list:
            return self._instantiate_items(list['extend']) + [list_item.ListItem(self._context, item, self) for item in list['items']]
        elif 'items' in list:
            return [list_item.ListItem(self._context, item, self) for item in list['items']]
        else:
            return []

    def _items(self):
        if not hasattr(self, '_instantiated_items'):
            self._instantiated_items = self._instantiate_items(self._instruction)

        return self._instantiated_items

    def _untouched(self):
        if not hasattr(self, '_touched'):
            return self._instruction

        untouched_item = next((item for item in self._items() if not hasattr(item, '_touched')), None)

        return untouched_item._instruction if untouched_item else False

    def items(self):
        self._touched = True

        return self._items()

    def length(self):
        self._touched = True

        return len(self._items())

    def optional_string_values(self):
        self._touched = True

        return [item.optional_string_value() for item in self._items()]

    def optional_values(self, loader):
        self._touched = True

        return [item.optional_value(loader) for item in self._items()]

    def parent(self):
        return self._parent or section.Section(self._context, self._instruction['parent'])

    def required_string_values(self):
        self._touched = True

        return [item.required_string_value() for item in self._items()]

    def required_values(self, loader):
        self._touched = True

        return [item.required_value(loader) for item in self._items()]

    def touch(self):
        self._touched = True

        for item in self._items():
            item._touched = True
